src/common/config_parser.cpp

```diff
@@ -72,997 +37,712 @@ const std::set<std::string> PATHS = {"model",
-  std::string devices = utils::Join(get<std::vector<std::string>>("devices"));
-  utils::Trim(devices);
+void ConfigParser::addOptionsGeneral(cli::CLIWrapper& cli) {
+  int defaultWorkspace = (mode_ == cli::mode::translation) ? 512 : 2048;
 
-  regex::regex pattern;
-  std::string help;
-  if(mode_ == ConfigMode::training && get<bool>("multi-node")) {
-    // valid strings: '0: 1 2', '0:1 2 1:2 3'
-    pattern = "( *[0-9]+ *: *[0-9]+( *[0-9]+)*)+";
-    help = "Supported format for multi-node setting: '0:0 1 2 3 1:0 1 2 3'";
-  } else {
-    // valid strings: '0', '0 1 2 3', '3 2 0 1'
-    pattern = "[0-9]+( *[0-9]+)*";
-    help = "Supported formats: '0 1 2 3'";
-  }
-
-  UTIL_THROW_IF2(!regex::regex_match(devices, pattern),
-                 "the argument '(" + devices
-                     + ")' for option '--devices' is invalid. " + help);
-}
+  cli.switchGroup("General options");
 
-void ConfigParser::addOptionsCommon(po::options_description& desc) {
-  int defaultWorkspace = (mode_ == ConfigMode::translating) ? 512 : 2048;
-
-  po::options_description general("General options", guess_terminal_width());
   // clang-format off
-  general.add_options()
-    ("config,c", po::value<std::vector<std::string>>()->multitoken(),
-     "Configuration file(s). If multiple, later overrides earlier.")
-    ("workspace,w", po::value<size_t>()->default_value(defaultWorkspace),
-      "Preallocate  arg  MB of work space")
-    ("log", po::value<std::string>(),
-     "Log training process information to file given by  arg")
-    ("log-level", po::value<std::string>()->default_value("info"),
-     "Set verbosity level of logging "
-     "(trace - debug - info - warn - err(or) - critical - off)")
-    ("quiet", po::value<bool>()->zero_tokens()->default_value(false),
-     "Suppress all logging to stderr. Logging to files still works")
-    ("quiet-translation", po::value<bool>()->zero_tokens()->default_value(false),
-     "Suppress logging for translation")
-    ("seed", po::value<size_t>()->default_value(0),
-     "Seed for all random number generators. 0 means initialize randomly")
-    ("clip-gemm", po::value<float>()->default_value(0.f),
-     "If not 0 clip GEMM input values to +/- arg")
-    ("interpolate-env-vars", po::value<bool>()->zero_tokens()->default_value(false),
-     "allow the use of environment variables in paths, of the form ${VAR_NAME}")
-    ("relative-paths", po::value<bool>()->zero_tokens()->default_value(false),
-     "All paths are relative to the config file location")
-    ("dump-config", po::value<bool>()->zero_tokens()->default_value(false),
-     "Dump current (modified) configuration to stdout and exit")
-    ("version", po::value<bool>()->zero_tokens()->default_value(false),
-      "Print version number and exit")
-    ("help,h", po::value<bool>()->zero_tokens()->default_value(false),
-      "Print this help message and exit")
-  ;
+  cli.add_nondefault<bool>("--version",
+      "Print version number and exit");
```

Comment:
I presume all of these are purely mechanical conversions without actual changes of strings or defaults?

Reply:
Yes. `add_nondefault` has been added only to preserve compatibility with the old code and hopefully it will be removed soon. Added TODO